src/channel/verifier_channel.rs

```diff
@@ -0,0 +1,78 @@
+use crate::channel::{Annotations, Channel, ChannelStates};
+use anyhow::Error;
+use ark_ff::Field;
+use std::io::Write;
+use std::vec::Vec;
+
+#[derive(Default)]
+pub struct VerifierChannel<F: Field> {
+    annotations: Annotations,
+    state: ChannelStates,
+    _field: std::marker::PhantomData<F>,
+}
+
+#[allow(dead_code)]
+impl<F: Field> Channel for VerifierChannel<F> {
+    type Field = F;
+
+    fn recv_felts(&mut self, n: usize) -> Result<Vec<Self::Field>, Error> {
+    }
+
+    fn recv_bytes(&mut self, n: usize) -> Result<Vec<u8>, Error> {
+    }
+
+    fn random_number(&mut self, bound: u64) -> u64 {
+    }
+
+    fn random_field(&mut self) -> Self::Field {
+    }
+
+    fn get_state(&self) -> &ChannelStates {
+        &self.state
+    }
+
+    fn get_state_mut(&mut self) -> &mut ChannelStates {
+        &mut self.state
+    }
+
+    fn get_annotations(&self) -> &Annotations {
+        &self.annotations
+    }
+
+    fn get_annotations_mut(&mut self) -> &mut Annotations {
+        &mut self.annotations
+    }
+}
+
+#[allow(dead_code)]
+impl<F: Field> VerifierChannel<F> {
+    pub fn get_and_send_random_number(&mut self, upper_bound: u64, annotation: &str) -> u64 {
+    }
+
+    pub fn get_and_send_random_field_element(&mut self, annotation: &str) -> F {
+    }
+
+    pub fn receive_commitment_hash(&mut self, annotation: &str) -> Vec<u8> {
+    }
+
+    pub fn recv_felt(&mut self, annotation: &str) -> F {
+    }
+
+    pub fn recv_data(&mut self, num_bytes: usize, annotation: &str) -> Vec<u8> {
+    }
+
+    pub fn dump_extra_annotations<W: Write>(&self, writer: &mut W) -> Result<(), std::io::Error> {
+        for annotation in &self.annotations.extra_annotations {
+            writeln!(writer, "{}", annotation)?;
+        }
+        Ok(())
+    }
+
+    pub fn annotate_extra_field_element(&mut self, field_element: F, annotation: &str) {
```

Comment:
Again this could be enum variants.

In the stone-prover the analogous method is used twice:

```c++
// Annotate previous FRI layer for fri-proof-splitting.
--
147 | channel_->AnnotateExtraFieldElement(query_results_[j], ElementDecommitAnnotation(query_loc));
148 | // Compute and annotate the x-inv of query j for fri-proof-splitting.
149 | FieldElement x_inv = basis.GetFieldElementAt(query_index).Inverse();
150 | channel_->AnnotateExtraFieldElement(x_inv, "xInv for index " + std::to_string(query_index));
```

So I would introduce enum variants, one for each. The fact that they are "flattened" to a string in the stone-prover is not very readable.

Reply:
Great! I strongly agree.
I was only thinking about how to adapt OOP impl. to Rust idioms, but I overlooked the fact that storing annotations as string continuously is a terrible design choice